### PR TITLE
Improve endpoint metadata debugging

### LINKED ISSUE #39792: Improve endpoint metadata debugging by overriding ToString
# Background and Motivation

It is useful to understand what behavior endpoint metadata is adding to an endpoint. A human-readable summary of metadata would make this much easier. It can be viewed when debugging, and in logs when metadata is logged.

https://github.com/dotnet/aspnetcore/issues/34604 is another potential use of human-readable endpoint metadata.

# Proposed API

I propose overriding ToString on metadata types. See changes in https://github.com/dotnet/aspnetcore/pull/35231 and https://github.com/dotnet/aspnetcore/pull/39753 as examples.

```
override Microsoft.AspNetCore.Routing.DataTokensMetadata.ToString() -> string!
override Microsoft.AspNetCore.Routing.EndpointNameMetadata.ToString() -> string!
override Microsoft.AspNetCore.Routing.HostAttribute.ToString() -> string!
override Microsoft.AspNetCore.Routing.HttpMethodMetadata.ToString() -> string!
override Microsoft.AspNetCore.Routing.RouteNameMetadata.ToString() -> string!
override Microsoft.AspNetCore.Routing.SuppressLinkGenerationMetadata.ToString() -> string!
override Microsoft.AspNetCore.Routing.SuppressMatchingMetadata.ToString() -> string!
```

# Alternative Designs

An alternative approach is to add an interface to get display information:

```csharp
public interface IEndpointDebugText
{
    string DebugText { get; }
}
```

I don't think this adds anything. We already have ToString that is pretty much for this situation.

`DebuggerDisplayAttribute` is not appropriate because there are situations where it is useful to get the debug value at runtime. e.g. logging. It is only useful if we provide debug text using something other than ToString.

# Risks

Someone uses endpoint metadata ToString to drive application behavior, and we change the ToString value. I don't think this is a valid concern. Someone could depend on text in exception messages and logging, and those change from time to time.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/39792

This PR:

* Adds `ToString` implementations to many types used as metadata on endpoints
* Adds `DebuggerDisplayAttribute` that uses the ToString
* Adds display and proxy attributes to `EndpointMetadataCollection`

Replaces https://github.com/dotnet/aspnetcore/pull/47163 by @SeanFarrow

Before:

image

After:

image

### CODE REVIEW

File: src/Http/Routing/src/SuppressLinkGenerationMetadata.cs [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
 {
     /// <summary>
     /// Gets a value indicating whether the assocated endpoint should be used for link generation.
     /// </summary>
     public bool SuppressLinkGeneration => true;
+
+    /// <inheritdoc/>
+    public override string ToString()
+    {
+        return $"SuppressLinkGeneration: {SuppressLinkGeneration}";
```
Comment (Tratcher) [inline src/Http/Routing/src/SuppressLinkGenerationMetadata.cs:23]: It won't ever be false. Same for SuppressMatchingMetadata 
```suggestion
        return "SuppressLinkGeneration: true";
```

File: src/Middleware/CORS/src/EnableCorsAttribute.cs [hunk 2]
```diff
@@ -29,4 +31,10 @@ public EnableCorsAttribute(string? policyName)
 
     /// <inheritdoc />
     public string? PolicyName { get; set; }
+
+    /// <inheritdoc/>>
+    public override string ToString()
+    {
+        return $"PolicyName: {PolicyName ?? "(null)"}";
```
Comment (Tratcher) [inline src/Middleware/CORS/src/EnableCorsAttribute.cs:38]:
```suggestion
        return $"CORS PolicyName: {PolicyName ?? "(null)"}";
```

Comment (Tratcher) [inline src/Middleware/CORS/src/EnableCorsAttribute.cs:38]:
```suggestion
        return $"CORS PolicyName: {PolicyName ?? "(null)"}";
```

File: src/Security/Authorization/Core/src/AuthorizeAttribute.cs [hunk 3]
```diff
@@ -39,4 +41,12 @@ public AuthorizeAttribute(string policy)
     /// Gets or sets a comma delimited list of schemes from which user information is constructed.
     /// </summary>
     public string? AuthenticationSchemes { get; set; }
+
+    /// <inheritdoc/>>
+    public override string ToString()
+    {
+        const string NullValue = "(null)";
+
+        return $"Policy: {Policy ?? NullValue}, Roles: {Roles ?? NullValue}, AuthenticationSchemes: {AuthenticationSchemes ?? NullValue}";
```
Comment (Tratcher) [inline src/Security/Authorization/Core/src/AuthorizeAttribute.cs:50]:
```suggestion
        return $"Authorization Policy: {Policy ?? NullValue}, Roles: {Roles ?? NullValue}, AuthenticationSchemes: {AuthenticationSchemes ?? NullValue}";
```

**Review by Tratcher — APPROVED** [review]

File: src/Security/Authorization/Core/src/AuthorizeAttribute.cs (hunk 3 above)
Comment (Tratcher) [reply src/Security/Authorization/Core/src/AuthorizeAttribute.cs:50]: Could this be more dynamic and only show the fields that are set? It's very common to only use one of these fields.
"Authorization Policy: MyAuthPolicy"
"Authorization Roles: Role1, Role2"
"AuthenticationSchemes: Bearer"

File: src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs [hunk 4]
```diff
@@ -215,4 +218,27 @@ public void Reset()
...
+
+            _collection = collection;
+        }
+
+        [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
+        public object[] Items
+        {
+            get
+            {
+                var items = new object[_collection.Count];
```
Comment (amcasey) [inline src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs:238]: I haven't made a debugger type proxy in a while - we're returning a copy to protect it from mutation in the locals/watch window?

File: src/Http/Http.Extensions/src/TagsAttribute.cs [hunk 5]
```diff
@@ -29,4 +31,10 @@ public TagsAttribute(params string[] tags)
     /// Gets the collection of tags associated with the endpoint.
     /// </summary>
     public IReadOnlyList<string> Tags { get; }
+
+    /// <inheritdoc/>>
+    public override string ToString()
+    {
+        return $"Tags: {string.Join(",", Tags)}";
```
Comment (amcasey) [inline src/Http/Http.Extensions/src/TagsAttribute.cs:38]: Would it be easier to read with `", "`?

File: src/Http/Routing/src/SuppressLinkGenerationMetadata.cs (hunk 1 above)
Comment (amcasey) [reply src/Http/Routing/src/SuppressLinkGenerationMetadata.cs:23]: I don't love hard-coding the value in two places.  Would it just be for the perf win?

File: src/Middleware/CORS/src/DisableCorsAttribute.cs [hunk 6]
```diff
@@ -1,12 +1,19 @@
...
 
 /// <inheritdoc />
 [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+[DebuggerDisplay("{ToString(),nq}")]
 public class DisableCorsAttribute : Attribute, IDisableCorsAttribute
 {
+    /// <inheritdoc/>>
+    public override string ToString()
+    {
+        return "DisableCors";
```
Comment (amcasey) [inline src/Middleware/CORS/src/DisableCorsAttribute.cs:17]: I don't quite understand the advantage of returning a constant string.

**Review by amcasey — APPROVED** [review]
Minor questions, but LGTM

File: src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs (hunk 4 above)
Comment (JamesNK) [reply src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs:238]: I don't know exactly. All the examples I saw in dotnet/runtime copied the collection, so I followed them.

File: src/Http/Http.Extensions/src/TagsAttribute.cs (hunk 5 above)
Comment (JamesNK) [reply src/Http/Http.Extensions/src/TagsAttribute.cs:38]: Comma + space indicates a new property. I didn't add spaces between items to make it more obvious they're in the same collection.

File: src/Middleware/CORS/src/DisableCorsAttribute.cs (hunk 6 above)
Comment (JamesNK) [reply src/Middleware/CORS/src/DisableCorsAttribute.cs:17]: I was on the fence about doing this for attributes with no properties.

The value is for debugging and potentially being displayed in logging and places like the developer exception page, e.g. https://github.com/dotnet/aspnetcore/issues/48208

We could keep returning the type name, but someone might wonder what properties are on `DisableCorsAttribute`. Returning a string like this makes it obvious there aren't any properties.

File: src/Http/Routing/src/SuppressLinkGenerationMetadata.cs (hunk 1 above)
Comment (JamesNK) [reply src/Http/Routing/src/SuppressLinkGenerationMetadata.cs:23]: Technically it should be `SuppressLinkGeneration: True` (capital T).

I don't care about perf here at all. I'd prefer to keep it understandable: here is a property name and here is its property value.

File: src/Security/Authorization/Core/src/AuthorizeAttribute.cs (hunk 3 above)
Comment (JamesNK) [reply src/Security/Authorization/Core/src/AuthorizeAttribute.cs:50]: That makes the implementation much more complicated.

I would want a helper method to do this. You're starting me down the path of writing a serializer 😱

Comment (Tratcher) [reply src/Security/Authorization/Core/src/AuthorizeAttribute.cs]: Ah, you wouldn't know anything about that...

File: src/Security/Authorization/Core/src/AuthorizeAttribute.cs [hunk 7]
```diff
@@ -39,4 +41,12 @@ public AuthorizeAttribute(string policy)
     /// Gets or sets a comma delimited list of schemes from which user information is constructed.
     /// </summary>
     public string? AuthenticationSchemes { get; set; }
+
+    /// <inheritdoc/>>
+    public override string ToString()
+    {
+        const string NullValue = "(null)";
+
+        return $"Authorize Policy: {Policy ?? NullValue}, Roles: {Roles ?? NullValue}, AuthenticationSchemes: {AuthenticationSchemes ?? NullValue}";
```
Comment (Tratcher) [inline src/Security/Authorization/Core/src/AuthorizeAttribute.cs:50]:
```suggestion
        var first = true;
        var builder = new StringBuilder("Authorize");
        if (Policy != null)
        {
              first = false;
              builder.Append(" Policy: ");
              builder.Append(Policy);
        }
        if (Roles != null)
        {
            if (first)
            {
                builder.Append(',');
                first = false;
            }
             builder.Append(" Roles: ");
             builder.Append(Roles);
        }
        if (AuthenticationSchemes != null)
        {
            if (first)
            {
                builder.Append(',');
                first = false;
            }
             builder.Append(" AuthenticationSchemes: ");
             builder.Append(AuthenticationSchemes);
        }
        return builder.ToString();
```

**Comment by JamesNK** [comment]
@Tratcher I added a helper method for building up a debug string for types. There are some allocations (the values array, string builder), but these methods are unlikely to ever be used on a hot path. Could avoid the values array in most cases by adding something like [TagList](https://github.com/dotnet/runtime/blob/ec67b315903e4716a5d043b4346f67baf77ca8a5/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/TagList.cs), but I'm not concerned about perf here so I skipped it.

What do you think?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**